### ISSUE TITLE
Help, regarding the issue of Qiskit aer using mpi to run programs on multiple nodes

### ISSUE THREAD

**Issue by a17513152253** [Qiskit/qiskit-aer#1955]
### What is the expected behavior?
Qiskit aer supports MPI. According to the tutorial, I can only run repeated results on one node. What if using MPI and having a large number of quantum programs run on multiple nodes? Can you give me a simple example? If you could help me, I would be extremely grateful.

**Comment by a17513152253** [comment]
For example, how can a 50qubit GHZ program run on 5 nodes? Can you give me guidance?

**Comment by doichanj** [comment]
Please refer to https://qiskit.org/ecosystem/aer/howtos/running_gpu.html for parallel execution. The document is for GPU but it is similar to run on CPU with multi-nodes.

By the way, if you run 50 qubits using statevector simulator you will need 512 TB of memory per node for 5 nodes (same as 4 nodes).

**Comment by a17513152253** [comment]
> 有关并行执行，请参阅 https://qiskit.org/ecosystem/aer/howtos/running_gpu.html。该文档适用于 GPU，但它类似于在具有多节点的 CPU 上运行。
> 
> 顺便说一下，如果您使用状态向量模拟器运行 50 个量子位，则每个节点需要 512 TB 的内存用于 5 个节点（与 4 个节点相同）。
Thank you for your reply,this is my code:
`import numpy as np
from mpi4py import MPI

# Import Qiskit
from qiskit import QuantumCircuit
from qiskit import Aer, transpile
#from qiskit.tools.visualization import plot_histogram, plot_state_city
#import qiskit.quantum_info as qi

comm = MPI.COMM_WORLD
#size = comm.Get_size()
rank = comm.Get_rank()
node_name = MPI.Get_processor_name()


# Create circuit
circ = QuantumCircuit(2)
circ.h(0)
circ.cx(0, 1)
circ.measure_all()

# Transpile for simulator
simulator = Aer.get_backend('aer_simulator')
simulator.set_options(device='CPU')
circ = transpile(circ, simulator)

# Run and get counts
result = simulator.run(circ).result()`
I will run this code on 2 nodes, one of which has 3 processes, which means there are 6 processes. But the results showed that it was{“11”：509，“”00：515} {“11”：493，“”00：531} {“11”：500，“”00：524} {“11”：491，“”00：515} {“11”：486，“”00：538}{“11”：514，“”00：510}
The result gives me the feeling that this code repeatedly runs independently on all processes on each node. That is to say, the code ran independently 6 times. But the effect I want to achieve is to split multiple qubits into fast, blocks run on one node, and finally the results on each node are combined to form the final answer. I'm not sure if my expression is accurate. If you understand, could you please explain it to me? Looking forward to your reply.

**Comment by a17513152253** [comment]
> Please refer to https://qiskit.org/ecosystem/aer/howtos/running_gpu.html for parallel execution. The document is for GPU but it is similar to run on CPU with multi-nodes.
> 
> By the way, if you run 50 qubits using statevector simulator you will need 512 TB of memory per node for 5 nodes (same as 4 nodes).

Thank you for your reply,this is my code:
`import numpy as np
from mpi4py import MPI


from qiskit import QuantumCircuit
from qiskit import Aer, transpile
#from qiskit.tools.visualization import plot_histogram, plot_state_city
#import qiskit.quantum_info as qi

comm = MPI.COMM_WORLD
#size = comm.Get_size()
rank = comm.Get_rank()
node_name = MPI.Get_processor_name()
circ = QuantumCircuit(2)
circ.h(0)
circ.cx(0, 1)
circ.measure_all()
simulator = Aer.get_backend('aer_simulator')
simulator.set_options(device='CPU')
circ = transpile(circ, simulator)
result = simulator.run(circ).result()`

I will run this code on 2 nodes, one of which has 3 processes, which means there are 6 processes. But the results showed that it was{“11”：509，“”00：515} {“11”：493，“”00：531} {“11”：500，“”00：524} {“11”：491，“”00：515} {“11”：486，“”00：538}{“11”：514，“”00：510}
The result gives me the feeling that this code repeatedly runs independently on all processes on each node. That is to say, the code ran independently 6 times. But the effect I want to achieve is to split multiple qubits into fast, blocks run on one node, and finally the results on each node are combined to form the final answer. I'm not sure if my expression is accurate. If you understand, could you please explain it to me? Looking forward to your reply.

**Comment by a17513152253** [comment]
> Please refer to https://qiskit.org/ecosystem/aer/howtos/running_gpu.html for parallel execution. The document is for GPU but it is similar to run on CPU with multi-nodes.
> 
> By the way, if you run 50 qubits using statevector simulator you will need 512 TB of memory per node for 5 nodes (same as 4 nodes).

Can you leave your email or other contact information? I want to communicate with you carefully. This question has been confusing me for too long.

**Comment by a17513152253** [comment]
That kind-hearted person saw my problem and if there is a solution, please help me. This problem has been confusing me for a long time。

**Comment by a17513152253** [comment]
Thank you for your reply,this is my code:
`import numpy as np
from mpi4py import MPI

from qiskit import QuantumCircuit
from qiskit import Aer, transpile
#from qiskit.tools.visualization import plot_histogram, plot_state_city
#import qiskit.quantum_info as qi

comm = MPI.COMM_WORLD
#size = comm.Get_size()
rank = comm.Get_rank()
node_name = MPI.Get_processor_name()
circ = QuantumCircuit(2)
circ.h(0)
circ.cx(0, 1)
circ.measure_all()
simulator = Aer.get_backend('aer_simulator')
simulator.set_options(device='CPU')
circ = transpile(circ, simulator)
result = simulator.run(circ).result()`

I will run this code on 2 nodes, one of which has 3 processes, which means there are 6 processes. But the results showed that it was{“11”：509，“”00：515} {“11”：493，“”00：531} {“11”：500，“”00：524} {“11”：491，“”00：515} {“11”：486，“”00：538}{“11”：514，“”00：510}
The result gives me the feeling that this code repeatedly runs independently on all processes on each node. That is to say, the code ran independently 6 times. But the effect I want to achieve is to split multiple qubits into fast, blocks run on one node, and finally the results on each node are combined to form the final answer. I'm not sure if my expression is accurate. If you understand, could you please explain it to me? Looking forward to your reply.



在 2023-10-11 10:03:19，"Jun Doi" ***@***.***> 写道：

Please refer to https://qiskit.org/ecosystem/aer/howtos/running_gpu.html for parallel execution. The document is for GPU but it is similar to run on CPU with multi-nodes.

By the way, if you run 50 qubits using statevector simulator you will need 512 TB of memory per node for 5 nodes (same as 4 nodes).

**Comment by a17513152253** [comment]
Can someone reply to me? I really need your help

**Comment by doichanj** [comment]
You can not parallelize 2 qubits circuits using MPI, but al least 3 qubits is required to parallelize. It is better to parallelize using `2^np` processes and you can parallelize nq qubits circuit where `nq > np + 2`

To parallelize the simulation, please include option `blocking_qubits` where nb qubits of statevector is allocated on each process and `nb < nq - np`
For example, simulating 25 qubits circuit using 4 processes (`nq = 25, np = 2`) `nb < 23` is set to the option `blocking_qubits = 22` 

Then the Python script is executed by using `mpirun` command (you do not need mpi4py if you do not process data parallelly in Python script)
The result of the simulation is distributed to each process so each process obtains the same result.

**Comment by doichanj** [comment]
Also MPI support is only available building from source code

**Comment by a17513152253** [comment]
Thank you again for your reply. What do you mean by saying 'Also MPI support is only available building from source code'? The qiskit-aer I am currently building is built according to the IBM official website. Do you need any additional operations?

If the installation of qiskit-aer does not require additional operations, could you give me a simple example?

I will be waiting for your reply all the time. This matter is very important to me and I really want to solve it



At 2023-10-13 10:19:21, "Jun Doi" ***@***.***> wrote:

You can not parallelize 2 qubits circuits using MPI, but al least 3 qubits is required to parallelize. It is better to parallelize using 2^np processes and you can parallelize nq qubits circuit where nq > np + 2

To parallelize the simulation, please include option blocking_qubits where nb qubits of statevector is allocated on each process and nb < nq - np
For example, simulating 25 qubits circuit using 4 processes (nq = 25, np = 2) nb < 23 is set to the option blocking_qubits = 22

Then the Python script is executed by using mpirun command (you do not need mpi4py if you do not process data parallelly in Python script)
The result of the simulation is distributed to each process so each process obtains the same result.

**Comment by doichanj** [comment]
Please refer to the document to build from source code with MPI support https://github.com/Qiskit/qiskit-aer/blob/main/CONTRIBUTING.md#building-with-mpi-support

**Comment by a17513152253** [comment]
Does referring to this paragraph mean that Qiskit-aer can now support multi node parallel simulation?













At 2023-10-13 10:42:46, "Jun Doi" ***@***.***> wrote:

Please refer to the document to build from source code with MPI support https://github.com/Qiskit/qiskit-aer/blob/main/CONTRIBUTING.md#building-with-mpi-support

**Comment by doichanj** [comment]
Yes Aer support MPI parallel simulation by building from source code. If you have installed by `pip install qiskit-aer` your Aer does not support MPI

**Comment by a17513152253** [comment]
I am ready to give it a try now, but I still want to take the liberty of asking if you could give me a simple example.













At 2023-10-13 10:52:25, "Jun Doi" ***@***.***> wrote:

Yes Aer support MPI parallel simulation by building from source code. If you have installed by pip install qiskit-aer your Aer does not support MPI

**Comment by a17513152253** [comment]
I now want to know how to write MPI programs that support qiskit aer parallel execution, and how the programs differ on personal laptops and supercomputers with multiple nodes. This program doesn't need to have any meaning, just give me an example.




I know it's presumptuous, but I have no other choice. If possible, I can pay for it.













At 2023-10-13 10:52:25, "Jun Doi" ***@***.***> wrote:

Yes Aer support MPI parallel simulation by building from source code. If you have installed by pip install qiskit-aer your Aer does not support MPI

**Comment by doichanj** [comment]
You do not have to write MPI codes if you do not execute pre- or post- processes in Python codes

Here is an example of QuantumVolume circuit to enable parallel execution
Saving this to qv.py and run the script by

mpirun -np 4 python qv.py

```
from qiskit import *
from qiskit.circuit import *
from qiskit.circuit.library import *
from qiskit.providers.aer import *

backend = AerSimulator(method='statevector')

shots = 1000
depth=10
qubits = 25
block_bits = 22

circuit = transpile(QuantumVolume(qubits, depth, seed=0),
                    backend=sim,
                    optimization_level=0)
circuit.measure_all()

result = backend.run(circuit,sim,shots=shots,seed_simulator=12345,blocking_qubits=block_bits).result()
if result_base.to_dict()['metadata']['mpi_rank'] == 0:
    print(result_base.to_dict()['results'][0]['time_taken'])
    print(sorted(result_base.to_dict()['results'][0]['data']['counts'].items(),key=lambda x:x[0]))
```

**Comment by a17513152253** [comment]
Thank you very much for your busy response. I know my request is very impolite, but the program example you provided is not universally applicable. The program I want is a large and simple quantum program, such as qubit=25, where several simple single qubit quantum gates are inserted into each circuit. Among them, two qubit gates (CONTs) are inserted into any two circuits, with about 10 CONTs.

We don't need to worry about what this circuit diagram is for, but now we do have such a quantum circuit. I hope this 25qubit quantum circuit can run on 4 nodes. The result of four nodes is the final result of the circuit diagram.
You also know that no matter how complex a quantum circuit is, it can be decomposed into single qubit gates and CONT quantum gates. Given this example, when we encounter other circuits, we only need to decompose them and compare them with the example you provided, which can be imitated.













At 2023-10-13 12:46:02, "Jun Doi" ***@***.***> wrote:

You do not have to write MPI codes if you do not execute pre- or post- processes in Python codes

Here is an example of QuantumVolume circuit to enable parallel execution
Saving this to qv.py and run the script by

mpirun -np 4 python qv.py

from qiskit import *
from qiskit.circuit import *
from qiskit.circuit.library import *
from qiskit.providers.aer import *

backend = AerSimulator(method='statevector')

shots = 1000
depth=10
qubits = 25
block_bits = 22

circuit = transpile(QuantumVolume(qubits, depth, seed=0),
                    backend=sim,
                    optimization_level=0)
circuit.measure_all()

result = backend.run(circuit,sim,shots=shots,seed_simulator=12345,blocking_qubits=block_bits).result()
if result_base.to_dict()['metadata']['mpi_rank'] == 0:
    print(result_base.to_dict()['results'][0]['time_taken'])
    print(sorted(result_base.to_dict()['results'][0]['data']['counts'].items(),key=lambda x:x[0]))

**Comment by doichanj** [comment]
We are also looking for the good example of useful large qubits circuits :)

If you want to learn how quantum circuit works, please try running some circuits on the Qiskit tutorials/learning. 
You can run parallelly in the same way I put above

**Comment by a17513152253** [comment]
I think you misunderstood my point. I don't want any useful examples now, just a large and simple quantum circuit running in parallel, without the need for this quantum circuit to be useful. For example:
H q[0]
H q[1]
H q[2]
H q[3]
H q[4]
H q[5]
H q[6]
H q[7]
H q[8]
H q[9]
H q[10]
H q[11]
H q[12]
H q[13]
H q[14]
H q[15]
H q[16]
H q[17]
H q[18]
H q[19]
H q[20]
H q[21]
H q[22]
H q[23]
H q[24]
T q[0]
S q[1]
X q[2]
Y q[3]
Z q[4]
S q[5]
CNOT q[12],q[7]
TOFFOLI q[4],q[5],q[0]
TOFFOLI q[14],q[15],q[8]
CNOT q[6],q[1]
CZ q[11],q[7]
CNOT q[18],q[15]
CNOT q[2],q[3]
CNOT q[21],q[17]
CNOT q[6],q[2]
CNOT q[24],q[20]
CNOT q[13],q[9]
TOFFOLI q[22],q[19],q[24]




You see, this is a 25qubit quantum circuit that I casually wrote. How can I make it run on 4 nodes?










At 2023-10-13 15:48:39, "Jun Doi" ***@***.***> wrote:

We are also looking for the good example of useful large qubits circuits :)

If you want to learn how quantum circuit works, please try running some circuits on the Qiskit tutorials/learning.
You can run parallelly in the same way I put above

**Comment by doichanj** [comment]
```
backend = AerSimulator(method='statevector')
result = backend.run(circuit,sim,shots=shots,seed_simulator=12345,blocking_qubits=block_bits).result()
```

you can pass your circuit as same as the example above

**Comment by a17513152253** [comment]
Thank you very much. I'll give it a try













At 2023-10-13 16:19:12, "Jun Doi" ***@***.***> wrote:

backend = AerSimulator(method='statevector')
result = backend.run(circuit,sim,shots=shots,seed_simulator=12345,blocking_qubits=block_bits).result()


you can pass your circuit as same as the example above

**Comment by a17513152253** [comment]
I follow your advice, this is my program:



import numpy as np

# Import Qiskit

from qiskit import QuantumCircuit

from qiskit import Aer, transpile

from qiskit.tools.visualization import plot_histogram, plot_state_city

import qiskit.quantum_info as qi




simulator = Aer.get_backend('aer_simulator')

# Create circuit

circ = QuantumCircuit(5)

circ.h(0)

circ.h(2)

circ.cx(0, 1)

circ.cx(2, 3)

circ.cx(3, 4)

circ.measure_all()




# Transpile for simulator

simulator = Aer.get_backend('aer_simulator_statevector')

circ = transpile(circ, simulator)




result = simulator.run(circ,simulator, shots=10, seed_simulator=12345,blocking_enable=True,blocking_qubits=2,memory=True).result()

memory = result.get_memory(circ)

print(memory)




From the program, it can also be seen that this is a 5qubit program, and the program is also very simple, block_ Bits=2.

Then I used mpirun - np 2 Python shiyan3. py, which is stored in shiyan3. py.
The results are as follows：

['00000', '00011', '11100', '00000', '00000', '11100', '11100', '11100', '00011', '00011']

['00000', '00011', '11100', '00000', '00000', '11100', '11100', '11100', '00011', '00011']



I still have three questions. Firstly, is the above operation the MPI parallelism of qiskit-aer? Secondly, why does the result appear twice? Does the result appear in both nodes。3、 How to know which qubit the node is running
When I use num_ Mpi_ When processes view how many processes there are in total, the result shows only one process. The result is still repeated twice. I want a node to run a qubit block,
Could you please help me find out where to go? Thank you very much


The current result is not what I want

At 2023-10-13 16:19:12, "Jun Doi" ***@***.***> wrote:

backend = AerSimulator(method='statevector')
result = backend.run(circuit,sim,shots=shots,seed_simulator=12345,blocking_qubits=block_bits).result()


you can pass your circuit as same as the example above

**Comment by doichanj** [comment]
1.  you can check if the simulation was parallelized or not to read `num_processes_per_experiments` in metadata of result
2. As I already mentioned above each process has the same result. And `print` function of Python script prints on each process, to avoid this, please read `mpi_rank` in metadata and write code to print only on process 0
3. qubits are not parallelized but statevector is distributed among processes (If you want to know details how to parallelize simulation, please refer to https://arxiv.org/abs/2102.02957)

**Comment by a17513152253** [comment]
I use mpirun - np 2 Python shiyan3. py to run the program in two processes. However, I checked num_ Processes_ Per_ Does experience=1 mean there is no parallelism. Why would this result be caused.




The second point，Do you mean to set conditions that only process 0 displays results?




Thirdly, I have been reading your paper, but now I prefer to successfully run the program in parallel.
















At 2023-10-13 17:30:27, "Jun Doi" ***@***.***> wrote:

you can check if the simulation was parallelized or not to read num_processes_per_experiments in metadata of result
As I already mentioned above each process has the same result. And print function of Python script prints on each process, to avoid this, please read mpi_rank in metadata and write code to print only on process 0
qubits are not parallelized but statevector is distributed among processes (If you want to know details how to parallelize simulation, please refer to https://arxiv.org/abs/2102.02957)

**Comment by a17513152253** [comment]
Good new week, let's continue with last week's topic,
I use mpirun - np 2 Python shiyan3. py to run the program in two processes. However, I checked num_ Processes_ Per_ Does experience=1 mean there is no parallelism. Why would this result be caused.
This is my program：



import numpy as np
# Import Qiskit
from qiskit.tools.visualization import plot_histogram, plot_state_city
import qiskit.quantum_info as qi


simulator = Aer.get_backend('aer_simulator')


# Create circuit
circ = QuantumCircuit(5)
circ.h(0)
circ.h(2)
circ.cx(0, 1)
circ.cx(2, 3)
circ.cx(3, 4)
circ.measure_all()


# Transpile for simulator
simulator = Aer.get_backend('aer_simulator_statevector')
circ = transpile(circ, simulator)


result = simulator.run(circ,simulator, shots=10, seed_simulator=12345,blocking_enable=True,blocking_qubits=2,memory=True).result()
dict = result.to_dict()
meta = dict['metadata']
myrank = meta['mpi_rank']
num = meta['num_mpi_processes']
#print(dict,meta,myrank)
print(num)
print(myrank)
memory = result.get_memory(circ)



Let me use mpirun - np 2 Python shiyan3. py, and the result is:

1

1

0

['00000', '00011', '11100', '00000', '00000', '11100', '11100', '11100', '00011', '00011']

0

['00000', '00011', '11100', '00000', '00000', '11100', '11100', '11100', '00011', '00011']



the num_mpi_processes=1.















At 2023-10-13 17:30:27, "Jun Doi" ***@***.***> wrote:

you can check if the simulation was parallelized or not to read num_processes_per_experiments in metadata of result
As I already mentioned above each process has the same result. And print function of Python script prints on each process, to avoid this, please read mpi_rank in metadata and write code to print only on process 0
qubits are not parallelized but statevector is distributed among processes (If you want to know details how to parallelize simulation, please refer to https://arxiv.org/abs/2102.02957)

**Comment by doichanj** [comment]
> Please refer to the document to build from source code with MPI support https://github.com/Qiskit/qiskit-aer/blob/main/CONTRIBUTING.md#building-with-mpi-support

**Comment by a17513152253** [comment]
Thank you very much for your guidance. Now that my goal has been achieved, I have implemented running on one node and two nodes using the program you provided me.



But I still have two questions. 1. A few days ago, you said, "To parallel the simulation, please include option blocking_qubits where nb qubits of stateactor are allocated on each process and nb<nq - np
For example, simulating 25 qubits circuit using 4 processes (nq=25, np=2) nb<23 is set to the option blocking_ Qubits=22“   np=2,Why do you use mpirun - np 4 Python qv. py instead of mpirun - np 2 Python qv. py



2. For example, 25 qubits, np=2, blocking_ Qubits=22, how can I know which part of qubits are running on which node
















在 2023-10-16 18:03:12，"Jun Doi" ***@***.***> 写道：

Please refer to the document to build from source code with MPI support https://github.com/Qiskit/qiskit-aer/blob/main/CONTRIBUTING.md#building-with-mpi-support

**Comment by a17513152253** [comment]
I'm sorry to bother you. I asked you about the parallel problems of qiskit-aer before, and you gave me answers. Now I have encountered a problem, which is still the program you gave me. But when I try to enlarge the qubit, say 29 or 30 bits, I get an error. The error results in the Primary job terminated normally, but 1 process returned

a non-zero exit code. Per user-direction, the job has been aborted.

--------------------------------------------------------------------------

--------------------------------------------------------------------------

mpirun noticed that process rank 1 with PID 28501 on node b06r4n18 exited on signal 9 (Killed).




Can you give me some answers? Why it went wrong.

The program is this:




import time

from qiskit import *

from qiskit.circuit import *

from qiskit.circuit.library import *

from qiskit.providers.aer import *




T1=time.time()




backend = AerSimulator(method='statevector')




shots = 100

depth=10

qubits = 29

block_bits = 23




circuit = transpile(QuantumVolume(qubits, depth, seed=0),

                    backend=backend,

                    optimization_level=0)

circuit.measure_all()




result = backend.run(circuit,backend,shots=shots,seed_simulator=12345,blocking_qubits=block_bits).result()

if result.to_dict()['metadata']['mpi_rank'] == 0:

    print(result.to_dict()['results'][0]['time_taken'])

    print(sorted(result.to_dict()['results'][0]['data']['counts'].items(),key=lambda x:x[0]))

    T2=time.time()

    print("运行时间："+str(T2 - T1)+"秒")

~                                             







在 2023-10-27 09:59:15，"Jun Doi" ***@***.***> 写道：

Closed #1955 as completed.
